Mathlib/Algebra/Category/CoalgebraCat/ComonEquivalence.lean

```diff
@@ -0,0 +1,197 @@
+the coalgebra axioms, but our actual `MonoidalCategory` instance on `CoalgebraCat` is
+constructed in `Mathlib.Algebra.Category.CoalgebraCat.Monoidal` to have better
+definitional equalities.
+
+-/
+universe v u
+
+namespace CoalgebraCat
+
+open CategoryTheory MonoidalCategory
+
+variable {R : Type u} [CommRing R]
+
+/-- An `R`-coalgebra is a comonoid object in the category of `R`-modules. -/
+@[simps] def toComonObj (X : CoalgebraCat R) : Comon_ (ModuleCat R) where
+  X := ModuleCat.of R X
+  counit := ModuleCat.ofHom Coalgebra.counit
+  comul := ModuleCat.ofHom Coalgebra.comul
+  counit_comul := by simpa only [ModuleCat.of_coe] using Coalgebra.rTensor_counit_comp_comul
+  comul_counit := by simpa only [ModuleCat.of_coe] using Coalgebra.lTensor_counit_comp_comul
+  comul_assoc := by
+    simp_rw [← Category.assoc, Iso.comp_inv_eq, ModuleCat.of_coe]
+    exact Coalgebra.coassoc.symm
+
+variable (R) in
+/-- The natural functor from `R`-coalgebras to comonoid objects in the category of `R`-modules. -/
+@[simps]
+def toComon : CoalgebraCat R ⥤ Comon_ (ModuleCat R) where
+  obj X := toComonObj X
+  map f :=
+    { hom := ModuleCat.ofHom f.1
+      hom_counit := f.1.counit_comp
+      hom_comul := f.1.map_comp_comul.symm }
+
+/-- A comonoid object in the category of `R`-modules has a natural comultiplication
+and counit. -/
+@[simps]
+instance ofComonObjCoalgebraStruct (X : Comon_ (ModuleCat R)) :
+    CoalgebraStruct R X.X where
+  comul := X.comul
+  counit := X.counit
+
+/-- A comonoid object in the category of `R`-modules has a natural `R`-coalgebra
+structure. -/
+def ofComonObj (X : Comon_ (ModuleCat R)) : CoalgebraCat R where
+  carrier := X.X
+  isCoalgebra :=
+    { ofComonObjCoalgebraStruct X with
+      coassoc := (LinearEquiv.eq_toLinearMap_symm_comp _ _).1 X.comul_assoc.symm
+      rTensor_counit_comp_comul := X.counit_comul
+      lTensor_counit_comp_comul := X.comul_counit }
+
+variable (R)
+
+/-- The natural functor from comonoid objects in the category of `R`-modules to `R`-coalgebras. -/
+def ofComon : Comon_ (ModuleCat R) ⥤ CoalgebraCat R where
+  obj X := ofComonObj X
+  map f :=
+    { toCoalgHom :=
+      { f.hom with
+        counit_comp := f.hom_counit
+        map_comp_comul := f.hom_comul.symm }}
+
+/-- The natural category equivalence between `R`-coalgebras and comonoid objects in the
+category of `R`-modules. -/
+@[simps]
+noncomputable def comonEquivalence : CoalgebraCat R ≌ Comon_ (ModuleCat R) where
+  functor := toComon R
+  inverse := ofComon R
+  unitIso := NatIso.ofComponents (fun _ => Iso.refl _) fun _ => by rfl
+  counitIso := NatIso.ofComponents (fun _ => Iso.refl _) fun _ => by rfl
+
+variable {R}
+
+/-- The monoidal category structure on the category of `R`-coalgebras induced by the
+equivalence with `Comon(R-Mod)`. This is just an auxiliary definition; the `MonoidalCategory`
+instance we make in `Mathlib.Algebra.Category.CoalgebraCat.Monoidal` will have better
+definitional equalities. -/
```

Comment:
Maybe link to the PR number of a follow-up PR, so that people will not be confused that this file does not yet exist in `master`.

Reply:
Done, thanks.